File: BAEKJOON(Python)/step 26/1202_G2.py
import heapq
import sys
input = sys.stdin.readline

N, K = map(int,input().split())
jewel = []
bag = []
for _ in range(N):
    M, V = map(int, input().split())
    heapq.heappush(jewel, (M, V))

for _ in range(K):
    heapq.heappush(bag, int(input()))

sum = 0
capable = []
for _ in range(K):
    bagc = heapq.heappop(bag)
    while jewel and bagc >= jewel[0][0]:
        w, v = heapq.heappop(jewel)
        heapq.heappush(capable, -v)
    if capable:
        sum -= heapq.heappop(capable)
    elif not jewel:
        break

print(sum)

File: BAEKJOON(Java)/step 26/_1202_G2.java
import java.io.*;
import java.util.*;

public class _1202_G2 {
  public static void main(String[] args) throws IOException {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    StringBuilder sb = new StringBuilder();

    String input[] = br.readLine().split(" ");
    int N = Integer.parseInt(input[0]), K = Integer.parseInt(input[1]);

    PriorityQueue<Jewelry> jewelries = new PriorityQueue<>();
    for (int i = 0; i < N; i++) {
      input = br.readLine().split(" ");
      jewelries.add(new Jewelry(Integer.parseInt(input[0]), Integer.parseInt(input[1])));
    }

    PriorityQueue<Integer> bags = new PriorityQueue<>();
    for (int i = 0; i < K; i++) {
      bags.add(Integer.parseInt(br.readLine()));
    }

    Long sum = 0L;
    PriorityQueue<Integer> capable = new PriorityQueue<>(Comparator.reverseOrder());
    for (int i = 0; i < K; i++) {
      int bag = bags.poll();
      while (!jewelries.isEmpty() && bag >= jewelries.peek().weight) {
        Jewelry jewelry = jewelries.poll();
        capable.add(jewelry.value);
      }
      if (!capable.isEmpty()) sum += capable.poll();
      else if (jewelries.isEmpty()) break;
    }

    sb.append(sum).append("\n");

    bw.write(sb.toString());
    bw.flush();
  }
}

class Jewelry implements Comparable<Jewelry> {
  public int weight, value;

  public Jewelry(int weight, int value) {
    this.weight = weight;
    this.value = value;
  }

  @Override
  public int compareTo(Jewelry jewelry) {
    if (jewelry.weight != this.weight)
      return Integer.compare(this.weight, jewelry.weight);
    return Integer.compare(jewelry.value, this.value);
  }
}